File: summary/go/20_bit_dp.go
package main

import (
	"fmt"
)

func bitDP(weights []int, values []int, capacity int) int {
	n := len(weights)
	dp := make([]int, 1<<n)

	for s := 1; s < (1 << n); s++ {
		for i := 0; i < n; i++ {
			if s&(1<<i) != 0 {
				prevSubset := s ^ (1 << i)
				if weights[i] <= capacity {
					dp[s] = max(dp[s], dp[prevSubset]+values[i])
				}
			}
		}
	}

	return dp[(1<<n)-1]
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	var n, capacity int
	fmt.Scan(&n, &capacity)
	weights := make([]int, n)
	values := make([]int, n)

	for i := 0; i < n; i++ {
		fmt.Scan(&weights[i], &values[i])
	}

	result := bitDP(weights, values, capacity)
	fmt.Printf("Maximum value using bit DP: %d\n", result)
}

File: summary/elixir/20_bit_dp.exs
defmodule BitDP do
  def solve(weights, values, capacity) do
    n = length(weights)
    dp = List.duplicate(0, 1 <<< n)

    Enum.reduce(1..(1 <<< n) - 1, dp, fn s, dp ->
      Enum.reduce(0..(n - 1), dp, fn i, dp ->
        if (s &&& (1 <<< i)) != 0 do
          prev_subset = s ^^^ (1 <<< i)
          if Enum.at(weights, i) <= capacity do
            List.update_at(dp, s, fn current_value ->
              max(current_value, Enum.at(dp, prev_subset) + Enum.at(values, i))
            end)
          else
            dp
          end
        else
          dp
        end
      end)
    end)
    |> List.last()
  end
end

# Example usage:
weights = IO.gets("Enter weights (space-separated): ") |> String.trim() |> String.split() |> Enum.map(&String.to_integer/1)
values = IO.gets("Enter values (space-separated): ") |> String.trim() |> String.split() |> Enum.map(&String.to_integer/1)
capacity = IO.gets("Enter capacity: ") |> String.trim() |> String.to_integer()

max_value = BitDP.solve(weights, values, capacity)
IO.puts("Maximum value using bit DP: #{max_value}")
